File: examples/crewai-qwen-demo/agent.py
"""
CrewAI + Qwen (DashScope) demo against vfs4Agent.

Prereq:
  1. In terminal A:   pnpm ingest ./examples/sample-docs && pnpm server
  2. In this folder:  python3.12 -m venv .venv && source .venv/bin/activate
                      pip install -r requirements.txt
                      cp .env.example .env  (edit if needed)
                      python agent.py "how do I authenticate with OAuth?"
"""

from __future__ import annotations

import os
import sys
import json
import requests
from typing import Type

from dotenv import load_dotenv
from pydantic import BaseModel, Field
from crewai import Agent, Task, Crew, LLM
from crewai.tools import BaseTool


load_dotenv()

VFS_URL = os.environ.get("VFS_SERVER_URL", "http://localhost:7801")
VFS_SESSION = os.environ.get("VFS_SESSION", "demo-session")
DASHSCOPE_KEY = os.environ["DASHSCOPE_API_KEY"]
DASHSCOPE_BASE = os.environ.get(
    "DASHSCOPE_BASE_URL", "https://dashscope.aliyuncs.com/compatible-mode/v1"
)
QWEN_MODEL = os.environ.get("QWEN_MODEL", "qwen-plus")


# ----------------------------------------------------------------------------
# VFS bridge tool
# ----------------------------------------------------------------------------


def _vfs_post(path: str, body: dict) -> dict:
    r = requests.post(
        f"{VFS_URL}{path}",
        json=body,
        headers={"X-VFS-Session": VFS_SESSION, "Content-Type": "application/json"},
        timeout=30,
    )
    r.raise_for_status()
    return r.json()


class VfsBashInput(BaseModel):
    command: str = Field(
        ...,
        description=(
            "A bash-like command to execute inside the documentation sandbox. "
            "Supports: ls, cat, head, tail, wc, find, tree, grep (-r/-i/-n/-l), "
            "awk, sed, sort, uniq. The docs are read-only under /docs."
        ),
    )


class VfsBashTool(BaseTool):
    name: str = "vfs_bash"
    description: str = (
        "Execute bash commands against a virtual filesystem of documentation. "
        "Example: `grep -rni 'oauth' /docs`, `cat /docs/auth/oauth.md`, "
        "`find /docs -name '*.md'`, `tree /docs`."
    )
    args_schema: Type[BaseModel] = VfsBashInput

    def _run(self, command: str) -> str:
        data = _vfs_post("/v1/bash", {"command": command})
        stdout = (data.get("stdout") or "").rstrip()
        stderr = (data.get("stderr") or "").rstrip()
        exit_code = data.get("exitCode", 0)
        parts: list[str] = []
        if stdout:
            parts.append(stdout)
        if stderr:
            parts.append(f"[stderr]\n{stderr}")
        if exit_code and not parts:
            parts.append(f"[exit {exit_code}]")
        out = "\n".join(parts) if parts else "(no output)"
        # Truncate to keep context manageable
        if len(out) > 6000:
            out = out[:6000] + "\n...[truncated]"
        return out


# ----------------------------------------------------------------------------
# Crew
# ----------------------------------------------------------------------------


def build_crew(question: str) -> Crew:
    llm = LLM(
        model=f"dashscope/{QWEN_MODEL}",
        base_url=DASHSCOPE_BASE,
        api_key=DASHSCOPE_KEY,
    )

    researcher = Agent(
        role="Docs Researcher",
        goal=(
            "Answer questions accurately by reading the project documentation "
            "exposed under /docs using bash commands."
        ),
        backstory=(
            "You are a careful technical writer. You never fabricate details. "
            "When a question comes in you first explore the docs layout with "
            "`tree /docs` or `find /docs -name '*.md'`, then grep for keywords, "
            "and finally `cat` the most relevant files to extract the answer."
        ),
        tools=[VfsBashTool()],
        llm=llm,
        allow_delegation=False,
        verbose=True,
        max_iter=8,
    )

    task = Task(
        description=(
            f"User question: {question}\n\n"
            "Answer using only evidence found in /docs. Cite file paths you read.\n"
            "Workflow hint:\n"
            "  1. `tree /docs` to see layout.\n"
            "  2. `grep -rni <keyword> /docs` to locate relevant files.\n"
            "  3. `cat <path>` to read the full doc.\n"
            "  4. Synthesize a concise answer with inline file citations."
        ),
        expected_output=(
            "A concise answer that directly addresses the question, followed by "
            "a `Sources:` list of file paths."
        ),
        agent=researcher,
    )

    return Crew(agents=[researcher], tasks=[task], verbose=True)


def main() -> None:
    question = (
        " ".join(sys.argv[1:]).strip()
        or "How do I authenticate with OAuth, and what is the refresh flow?"
    )
    print(f"[demo] question: {question}")
    print(f"[demo] vfs:      {VFS_URL}")
    print(f"[demo] model:    {QWEN_MODEL}")

    # Sanity ping
    try:
        ping = _vfs_post("/v1/bash", {"command": "ls /docs"})
        print(f"[demo] vfs ping: {json.dumps(ping)[:200]}")
    except Exception as e:
        print(f"[demo] ERROR: cannot reach vfs-server at {VFS_URL}: {e}")
        sys.exit(2)

    crew = build_crew(question)
    result = crew.kickoff()
    print("\n" + "=" * 60)
    print("FINAL ANSWER")
    print("=" * 60)
    print(result)


if __name__ == "__main__":
    main()

File: examples/claude-agent-sdk-demo/agent.ts
/**
 * Claude Agent SDK demo — vfs4Agent
 *
 * Uses @anthropic-ai/sdk directly (the raw tool_use API) because the
 * Claude Agent SDK's built-in `Bash` tool runs on the host, not against
 * our /vfs filesystem. We declare a single custom `bash` tool and route
 * every invocation to vfs4Agent's HTTP /v1/bash endpoint.
 *
 * Run:
 *   pnpm server                                 # terminal A (docker on macOS)
 *   # or: python examples/_mock/mock_vfs_server.py
 *   cd examples/claude-agent-sdk-demo
 *   pnpm install
 *   ANTHROPIC_API_KEY=sk-ant-... pnpm start "OAuth refresh flow?"
 */

import "dotenv/config";
import Anthropic from "@anthropic-ai/sdk";

const VFS_URL = process.env.VFS_SERVER_URL ?? "http://localhost:7801";
const VFS_TOKEN = process.env.VFS_SESSION_TOKEN;
const MODEL = process.env.ANTHROPIC_MODEL ?? "claude-sonnet-4-5";

// ── VFS bridge ──────────────────────────────────────────────────────────
async function vfsBash(command: string): Promise<{
  stdout: string;
  stderr: string;
  exitCode: number;
}> {
  const res = await fetch(`${VFS_URL}/v1/bash`, {
    method: "POST",
    headers: {
      "Content-Type": "application/json",
      ...(VFS_TOKEN ? { "x-vfs-session": VFS_TOKEN } : {}),
    },
    body: JSON.stringify({ command }),
  });
  if (!res.ok) throw new Error(`vfs /v1/bash failed: ${res.status}`);
  return res.json() as Promise<{ stdout: string; stderr: string; exitCode: number }>;
}

function toolResultString(r: { stdout: string; stderr: string; exitCode: number }): string {
  const parts: string[] = [];
  if (r.stdout.trim()) parts.push(r.stdout.trimEnd());
  if (r.stderr.trim()) parts.push(`[stderr]\n${r.stderr.trimEnd()}`);
  if (r.exitCode && parts.length === 0) parts.push(`[exit ${r.exitCode}]`);
  const out = parts.join("\n") || "(no output)";
  return out.length > 6000 ? `${out.slice(0, 6000)}\n...[truncated]` : out;
}

// ── Tool definition (Anthropic schema) ──────────────────────────────────
const TOOLS: Anthropic.Messages.Tool[] = [
  {
    name: "bash",
    description:
      "Execute a bash command in the /vfs documentation sandbox. " +
      "Use ls, cat, grep, find, tree, head, tail, wc etc. Read-only under /vfs.",
    input_schema: {
      type: "object",
      properties: {
        command: { type: "string", description: "Bash command to run" },
      },
      required: ["command"],
    },
  },
];

const SYSTEM = `You are a documentation assistant with bash access to a \
read-only documentation tree mounted at /vfs.

Workflow: (1) \`ls /vfs\` or \`tree /vfs\` for layout. \
(2) \`grep -rni <keyword> /vfs\` to locate files. \
(3) \`cat /vfs/path\` to read. (4) Answer concisely and cite file paths.`;

// ── Agent loop ──────────────────────────────────────────────────────────
async function ask(question: string): Promise<string> {
  const client = new Anthropic();
  const messages: Anthropic.Messages.MessageParam[] = [
    { role: "user", content: question },
  ];

  for (let turn = 0; turn < 12; turn++) {
    const resp = await client.messages.create({
      model: MODEL,
      max_tokens: 4096,
      system: SYSTEM,
      tools: TOOLS,
      messages,
    });

    if (resp.stop_reason === "end_turn") {
      const text = resp.content
        .filter((b): b is Anthropic.Messages.TextBlock => b.type === "text")
        .map((b) => b.text)
        .join("\n");
      return text;
    }

    messages.push({ role: "assistant", content: resp.content });

    // Execute every tool_use block, batch results as a single user message.
    const results: Anthropic.Messages.ToolResultBlockParam[] = [];
    for (const block of resp.content) {
      if (block.type !== "tool_use") continue;
      if (block.name !== "bash") {
        results.push({
          type: "tool_result",
          tool_use_id: block.id,
          content: `unsupported tool: ${block.name}`,
          is_error: true,
        });
        continue;
      }
      const { command } = block.input as { command: string };
      console.log(`  ⚙  ${command}`);
      const r = await vfsBash(command);
      console.log(`  →  exit=${r.exitCode}`);
      results.push({
        type: "tool_result",
        tool_use_id: block.id,
        content: toolResultString(r),
        is_error: r.exitCode !== 0,
      });
    }
    messages.push({ role: "user", content: results });
  }

  return "[agent] max turns reached";
}

// ── Entry ───────────────────────────────────────────────────────────────
const question =
  process.argv.slice(2).join(" ").trim() ||
  "How do I authenticate with OAuth and what is the refresh flow?";

ask(question).then((answer) => {
  console.log("\n" + "=".repeat(60));
  console.log("FINAL ANSWER");
  console.log("=".repeat(60));
  console.log(answer);
});
